lib/spack/spack/build_systems/oneapi.py

```diff
@@ -195,6 +196,47 @@ def libs(self):
         # for v2_layout all libraries are in the top level, v1 sometimes put them in intel64
         return find_libraries("*", root=self.component_prefix.lib, recursive=not self.v2_layout)
 
+    @property
+    def openmp_libs(self):
+        """Supply LibraryList for linking OpenMP"""
+
+        if "%intel" in self.spec:
+            # NB: Hunting down explicit library files may be the Spack way of
+            # doing things, but be aware that "{icc|ifort} --help openmp"
+            # steers us towards options instead: -qopenmp-link={dynamic,static}
+
+            omp_libname = "libiomp5"
+            omp_libs = find_libraries(
+                omp_libname,
+                root=self.component_lib_dir("compiler"),
+                shared=("+shared" in self.spec),
+            )
+            # Note about search root here: For MKL, the directory
+            # "$MKLROOT/../compiler" will be present even for an MKL-only
+            # product installation (as opposed to one being ghosted via
+            # packages.yaml), specificially to provide the 'iomp5' libs.
+
+        elif "%gcc" in self.spec:
+            omp_libname = "libgomp"
+            with self.compiler.compiler_environment():
+                omp_lib_path = Executable(self.compiler.cc)(
+                    "--print-file-name", f"{omp_libname}.{dso_suffix}", output=str
+                )
+            omp_libs = LibraryList(omp_lib_path.strip())
+
+        elif "%clang" in self.spec:
+            omp_libname = "libomp"
+            with self.compiler.compiler_environment():
+                omp_lib_path = Executable(self.compiler.cc)(
+                    "--print-file-name", f"{omp_libname}.{dso_suffix}", output=str
+                )
+            omp_libs = LibraryList(omp_lib_path.strip())
```

Comment:
Let me take this opportunity to mention this PR, which changes the way OpenMP is added to intel packages: https://github.com/spack/spack/pull/38719. Instead of using library names/paths, I think this should be using the OpenMP flags, at least for clang. The original issue (https://github.com/spack/spack/issues/38509) has more details about what works and what doesn't. The issue was reported for `intel-mkl` but I'm guessing the same issue would be there for `intel-oneapi-mkl`.

Reply:
I had a look also at [intel compiler user manual](https://www.intel.com/content/www/us/en/docs/cpp-compiler/developer-guide-reference/2021-10/add-openmp-support.html) and it looks a bit trickier (see notes about macOS and LD_LIBRARY_PATH in [`-qopenmp`/`-fopenmp` flags](https://www.intel.com/content/www/us/en/docs/cpp-compiler/developer-guide-reference/2021-10/qopenmp-qopenmp.html#GUID-4AC3D696-F9A5-4064-A34A-A948300B80C7))

In addition to this, I'm trying to understand if from the package we can access the compiler property `openmp_flags`.

Reply:
> I would change gcc too, since that's what they report also in [their manual](https://gcc.gnu.org/onlinedocs/gcc-13.2.0/libgomp/Enabling-OpenMP.html) (I checked a couple of latest versions starting from gcc@10).

I realize now the one possible downside of using this approach is that when mixing compilers you may also end up mixing OpenMP runtimes. If an intel package is built with e.g. icc it would/should be using `libiomp`, but a dependent package built with e.g. GCC could end up with `libgomp` (since it doesn't explicitly specify which runtime to use, but just says `-fopenmp`). I suppose this might be one reason why things were set up the way they were.

However, the current state of this is still an improvement and I don't have a suggestion on how to better model this in spack.